<a href="https://colab.research.google.com/github/OTD-Aregbesola/Bank-personal-loan-modelling/blob/main/Bank_personal_loan_modelling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Bank personal loan modelling

This notebook provides an approach to feature selection for your target data when developing a model.

Data from Kaggle
https://www.kaggle.com/datasets/teertha/personal-loan-modeling?resource=download

In [21]:
# load the data
import pandas as pd
import numpy as np
from statsmodels.formula.api import ols
from scipy.stats import pearsonr

df = pd.read_csv('/content/Bank_Personal_Loan_Modelling.csv')

# Data Preprocessing

In [22]:
df.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


In [24]:
df.describe().T

count          mean          std     min       25%  \
ID                  5000.0   2500.500000  1443.520003     1.0   1250.75   
Age                 5000.0     45.338400    11.463166    23.0     35.00   
Experience          5000.0     20.104600    11.467954    -3.0     10.00   
Income              5000.0     73.774200    46.033729     8.0     39.00   
ZIP Code            5000.0  93152.503000  2121.852197  9307.0  91911.00   
Family              5000.0      2.396400     1.147663     1.0      1.00   
CCAvg               5000.0      1.937938     1.747659     0.0      0.70   
Education           5000.0      1.881000     0.839869     1.0      1.00   
Mortgage            5000.0     56.498800   101.713802     0.0      0.00   
Personal Loan       5000.0      0.096000     0.294621     0.0      0.00   
Securities Account  5000.0      0.104400     0.305809     0.0      0.00   
CD Account          5000.0      0.060400     0.238250     0.0      0.00   
Online              5000.0      0.596800     0.490589     0.0      0.00   
CreditCard          5000.0      0.294000     0.455637     0.0      0.00   

                        50%       75%      max  
ID                   2500.5   3750.25   5000.0  
Age                    45.0     55.00     67.0  
Experience             20.0     30.00     43.0  
Income                 64.0     98.00    224.0  
ZIP Code            93437.0  94608.00  96651.0  
Family                  2.0      3.00      4.0  
CCAvg                   1.5      2.50     10.0  
Education               2.0      3.00      3.0  
Mortgage                0.0    101.00    635.0  
Personal Loan           0.0      0.00      1.0  
Securities Account      0.0      0.00      1.0  
CD Account              0.0      0.00      1.0  
Online                  1.0      1.00      1.0  
CreditCard              0.0      1.00      1.0

In [25]:
#convert the columns name to lowercase and replace the white spaces with underscore
df.columns=  [col.lower().replace(' ','_') for col in df.columns]

df.head()

id  age  experience  income  zip_code  family  ccavg  education  mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   personal_loan  securities_account  cd_account  online  creditcard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [26]:
# drop the irrelevant columns
df.drop(columns = ['zip_code','id'], inplace = True)

#Data Analysis

This section demonstrates the use of the Ordinary Least Squares (OLS) method to construct the model. The model is first built before utilising the Pearson correlation p-value to select the appropriate features for the target variable.

In [27]:
#Define the X and y values
y_data = 'personal_loan'

X_data = [col for col in df.columns if col != 'personal_loan']

formular = y_data +'~'+'+'.join(X_data)

In [28]:
model_1 = ols(formula = formular, data = df)

fitted_1 = model_1.fit()

print(fitted_1.summary())


                            OLS Regression Results                            
Dep. Variable:          personal_loan   R-squared:                       0.386
Model:                            OLS   Adj. R-squared:                  0.385
Method:                 Least Squares   F-statistic:                     285.3
Date:                Sat, 04 Jan 2025   Prob (F-statistic):               0.00
Time:                        11:30:25   Log-Likelihood:                 236.48
No. Observations:                5000   AIC:                            -449.0
Df Residuals:                    4988   BIC:                            -370.8
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept             -0.2324      0

In the model, the R-squared value is 0.386, indicating that the features explain only 38.6% of the target variable. This suggests poor model performance, and there is evidence of multicollinearity in some of the feature data.

In [29]:
#check the correlation between the target and features variable
corrs = df.corr()['personal_loan'].sort_values(ascending = False)

corrs

personal_loan         1.000000
income                0.502462
ccavg                 0.366889
cd_account            0.316355
mortgage              0.142095
education             0.136722
family                0.061367
securities_account    0.021954
online                0.006278
creditcard            0.002802
experience           -0.007413
age                  -0.007726
Name: personal_loan, dtype: float64

In [30]:
# calculate the p_value
dict_p = {}

column_titles = [ col for col in corrs.index if col != 'personal_loan']


for col in column_titles:
  p_value = round(pearsonr(df[col], df['personal_loan'])[1],6)
  dict_p[col] = {'correlation_coefficient': corrs[col], 'p_value': p_value}



df_fr = pd.DataFrame(dict_p).T
df_fr_sorted = df_fr.sort_values('p_value')
df_fr_sorted = df_fr_sorted[df_fr_sorted['p_value'] < 0.05]

df_fr_sorted

correlation_coefficient   p_value
income                     0.502462  0.000000
ccavg                      0.366889  0.000000
cd_account                 0.316355  0.000000
mortgage                   0.142095  0.000000
education                  0.136722  0.000000
family                     0.061367  0.000014

In [31]:
# Check for multicollinerity among the new feature
new_feature = [col for col in df_fr_sorted.index]


df[new_feature].corr()

income     ccavg  cd_account  mortgage  education    family
income      1.000000  0.645984    0.169738  0.206806  -0.187524 -0.157501
ccavg       0.645984  1.000000    0.136534  0.109905  -0.136124 -0.109275
cd_account  0.169738  0.136534    1.000000  0.089311   0.013934  0.014110
mortgage    0.206806  0.109905    0.089311  1.000000  -0.033327 -0.020445
education  -0.187524 -0.136124    0.013934 -0.033327   1.000000  0.064929
family     -0.157501 -0.109275    0.014110 -0.020445   0.064929  1.000000

In [32]:
# build a model with the new features
y_name = y_data

X_name = new_feature

formular = y_data + '~' + '+'.join(new_feature)

model_ = ols(formula = formular, data = df)

model_2 =model_.fit()

print(model_2.summary())


                            OLS Regression Results                            
Dep. Variable:          personal_loan   R-squared:                       0.377
Model:                            OLS   Adj. R-squared:                  0.376
Method:                 Least Squares   F-statistic:                     503.2
Date:                Sat, 04 Jan 2025   Prob (F-statistic):               0.00
Time:                        11:35:29   Log-Likelihood:                 198.54
No. Observations:                5000   AIC:                            -383.1
Df Residuals:                    4993   BIC:                            -337.5
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.3995      0.013    -30.760      0.0

Conclusion
The R-squared value for the model decreased instead of increasing, indicating that the selected features are not suitable for predicting the target variable.